### PR TITLE
Support Rain Cloud DNS Provider

### PR DESCRIPTION
**Pull request by Cikaros**
The individual has been tested and is available normally~

### CODE REVIEW

**Comment by Cikaros** [comment]
// TODO(ldez) based on the API style, this structure is probably wrong.
There's no problem with the JSON field here, because I've tried calling the API and the property is determined to be a big hump and not a small hump.
image

**Comment by Cikaros** [comment]
I'm sorry for the irregularity of what I wrote, I rarely write go

**Comment by ldez** [comment]
Please stop merging, instead rebase your local branch.

**Comment by ldez** [comment]
To fix your local branch quickly:
```bash
git switch master
git branch -D feature/rain-cloud
git switch feature/rain-cloud
```

**Comment by ldez** [comment]
Please stop with that.

**Comment by Cikaros** [comment]
Okay , I`m so sorry

**Comment by ldez** [comment]
I'm waiting for:

[ ] be able to do: (and put the output of this command to a comment in your PR)
```
rm -rf .lego

RAINYUN_API_KEY="xxxxxxxxxxxxxxxxxxxxx" \
./dist/lego -m your@email.com --dns rainyun -d *.example.com -d example.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
```

**Note the wildcard domain is important.**

You should replace `example.com` with your domain.
You should replace `your@email.com` with your email.

The 2 domains with the same LTD are a requirement.

**Comment by Cikaros** [comment]
The interface is currently functioning normally; however, the inspection process is very lengthy, and I am unsure of the reasons behind this. 

```shell
2024/11/19 10:23:08 [INFO] [cikaros.cn] acme: Obtaining bundled SAN certificate
2024/11/19 10:23:09 [INFO] [cikaros.cn] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/14979299033
2024/11/19 10:23:09 [INFO] [cikaros.cn] acme: Could not find solver for: tls-alpn-01
2024/11/19 10:23:09 [INFO] [cikaros.cn] acme: Could not find solver for: http-01
2024/11/19 10:23:09 [INFO] [cikaros.cn] acme: use dns-01 solver
2024/11/19 10:23:09 [INFO] [cikaros.cn] acme: Preparing to solve DNS-01
2024/11/19 10:23:10 [INFO] [cikaros.cn] acme: Trying to solve DNS-01
2024/11/19 10:23:11 [INFO] [cikaros.cn] acme: Checking DNS record propagation. [nameservers=100.73.255.254:53]
2024/11/19 10:23:13 [INFO] Wait for propagation [timeout: 2m0s, interval: 2s]
2024/11/19 10:23:13 [INFO] [cikaros.cn] acme: Waiting for DNS record propagation.
2024/11/19 10:25:13 [INFO] [cikaros.cn] acme: Waiting for DNS record propagation.
2024/11/19 10:25:15 [INFO] [cikaros.cn] acme: Cleaning DNS-01 challenge
2024/11/19 10:28:21 [INFO] Deactivating auth: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/14979299033
2024/11/19 10:28:21 Could not obtain certificates:
        error: one or more domains had a problem:
[cikaros.cn] propagation: time limit exceeded: last error: authoritative nameservers: NS ns3.myhostadmin.net.:53 returned NXDOMAIN for _acme-challenge.cikaros.cn.
Exiting.
 ```
 
image

**Comment by ldez** [comment]
> [cikaros.cn] propagation: time limit exceeded: last error: authoritative nameservers: NS ns3.myhostadmin.net.:53 returned NXDOMAIN for _acme-challenge.cikaros.cn.

Several possibilities:
1. a slow propagation of the DNS entries
2. your authoritative nameserver has a problem
3. your domain is misconfigured

---

For the slow propagation:
```
rm -rf .lego

RAINYUN_PROPAGATION_TIMEOUT=120 \
RAINYUN_POLLING_INTERVAL=30 \
RAINYUN_API_KEY="xxxxxxxxxxxxxxxxxxxxx" \
./dist/lego -m your@email.com --dns rainyun -d *.example.com -d example.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
```

---

For authoritative nameserver:
```
drill @ns3.myhostadmin.net. _acme-challenge.cikaros.cn
```

```
rm -rf .lego

RAINYUN_API_KEY="xxxxxxxxxxxxxxxxxxxxx" \
./dist/lego -m your@email.com --dns rainyun -d *.example.com -d example.com -s https://acme-staging-v02.api.letsencrypt.org/directory --dns-resolvers 1.1.1.1 run
```

---

For the domain misconfiguration:

Maybe your problem is related to your CNAME.

**Comment by ldez** [comment]
You are continuing with your merge...

**Comment by Cikaros** [comment]
image
The reason for writing additional structs is that the HTTP API does not allow the acceptance of extra parameters, such as providing record_id=0 when adding a new record.

**Comment by ldez** [comment]
You don't understand what the struct tags are doing.

`omitempty` remove all the empty elements.

**Comment by Cikaros** [comment]
but it`s not nil
image

**Comment by Cikaros** [comment]
I am very sorry, it was my mistake.

**Comment by Cikaros** [comment]
image

now, it`s ok

**Comment by ldez** [comment]
What did you change? is it the propagation configuration?

can you give the text of the output of the command (not a screenshot)?

**Comment by Cikaros** [comment]
image
image
I believe there are still issues here; the domain name provided is the one that is about to be applied for. May I ask how I can obtain the main domain in Present and CleanUp?

**Comment by ldez** [comment]
Sorry, but I don't understand what you are saying, can you rephrase, please?

**Comment by ldez** [comment]
Why are you using `RAINYUN_TTL` = `300` ?

**Comment by Cikaros** [comment]
The parameter '-d <domain>' is provided for the domain that is about to be applied for, am I correct in my understanding? For example: If I need to apply for a certificate for 'www.cikaros.cn', then what is actually provided here is a subdomain.

**Comment by Cikaros** [comment]
There is no special significance; it is simply because I found that the default TTL in the DNS panel is 300.

**Comment by Cikaros** [comment]
Do you still remember the findDomainID() function? It requires the main domain name to find the corresponding ID.

**Comment by ldez** [comment]
You just need to use `-d www.cikaros.cn` but it's mainly useless because you should have a certificate for the wildcard domain `-d *.cikaros.cn`.

You should follow the instructions: https://github.com/go-acme/lego/pull/2354#issuecomment-2483241045

**Comment by ldez** [comment]
I will fix the find domain ID.

**Comment by Cikaros** [comment]
Alright, I will test it once more.

**Comment by Cikaros** [comment]
As requested, I have passed the test. But I've appended a new environment variable RAINYUN_PROPAGATION_TIMEOUT=120

**Comment by ldez** [comment]
OK, I changed the default propagation, and now it's 120.

Can you give me the text of the output of the run of the command: https://github.com/go-acme/lego/pull/2354#issuecomment-2483241045

**Comment by Cikaros** [comment]
```log
2024/11/19 11:45:41 [INFO] [*.cikaros.cn, cikaros.cn] acme: Obtaining bundled SAN certificate
2024/11/19 11:45:43 [INFO] [*.cikaros.cn] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/14980097623
2024/11/19 11:45:43 [INFO] [cikaros.cn] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/14980192463
2024/11/19 11:45:43 [INFO] [cikaros.cn] acme: authorization already valid; skipping challenge
2024/11/19 11:45:43 [INFO] [*.cikaros.cn] acme: use dns-01 solver
2024/11/19 11:45:43 [INFO] [*.cikaros.cn] acme: Preparing to solve DNS-01
2024/11/19 11:45:45 [INFO] [*.cikaros.cn] acme: Trying to solve DNS-01
2024/11/19 11:45:45 [INFO] [*.cikaros.cn] acme: Checking DNS record propagation. [nameservers=100.73.255.254:53]
2024/11/19 11:45:47 [INFO] Wait for propagation [timeout: 2m0s, interval: 2s]
2024/11/19 11:45:47 [INFO] [*.cikaros.cn] acme: Waiting for DNS record propagation.
2024/11/19 11:45:49 [INFO] [*.cikaros.cn] acme: Waiting for DNS record propagation.
2024/11/19 11:45:52 [INFO] [*.cikaros.cn] acme: Waiting for DNS record propagation.
2024/11/19 11:45:54 [INFO] [*.cikaros.cn] acme: Waiting for DNS record propagation.
2024/11/19 11:45:56 [INFO] [*.cikaros.cn] acme: Waiting for DNS record propagation.
2024/11/19 11:45:58 [INFO] [*.cikaros.cn] acme: Waiting for DNS record propagation.
2024/11/19 11:46:00 [INFO] [*.cikaros.cn] acme: Waiting for DNS record propagation.
2024/11/19 11:46:02 [INFO] [*.cikaros.cn] acme: Waiting for DNS record propagation.
2024/11/19 11:46:05 [INFO] [*.cikaros.cn] acme: Waiting for DNS record propagation.
2024/11/19 11:46:07 [INFO] [*.cikaros.cn] acme: Waiting for DNS record propagation.
2024/11/19 11:46:09 [INFO] [*.cikaros.cn] acme: Waiting for DNS record propagation.
2024/11/19 11:46:11 [INFO] [*.cikaros.cn] acme: Waiting for DNS record propagation.
2024/11/19 11:46:13 [INFO] [*.cikaros.cn] acme: Waiting for DNS record propagation.
2024/11/19 11:46:16 [INFO] [*.cikaros.cn] acme: Waiting for DNS record propagation.
2024/11/19 11:46:18 [INFO] [*.cikaros.cn] acme: Waiting for DNS record propagation.
2024/11/19 11:46:20 [INFO] [*.cikaros.cn] acme: Waiting for DNS record propagation.
2024/11/19 11:46:22 [INFO] [*.cikaros.cn] acme: Waiting for DNS record propagation.
2024/11/19 11:46:25 [INFO] [*.cikaros.cn] acme: Waiting for DNS record propagation.
2024/11/19 11:46:42 [INFO] [*.cikaros.cn] The server validated our request
2024/11/19 11:46:42 [INFO] [*.cikaros.cn] acme: Cleaning DNS-01 challenge
2024/11/19 11:46:43 [INFO] [*.cikaros.cn, cikaros.cn] acme: Validations succeeded; requesting certificates
2024/11/19 11:46:43 [INFO] Wait for certificate [timeout: 30s, interval: 500ms]
2024/11/19 11:46:49 [INFO] [*.cikaros.cn] Server responded with a certificate.
```

**Comment by ldez** [comment]
So everything seems OK now.

**Comment by Cikaros** [comment]
Yes, everything is fine. Thanks for the guidance~

**Comment by Cikaros** [comment]
Can it be merged?

**Review by ldez — APPROVED** [review]
LGTM
